src/lib.rs

```diff
@@ -1,101 +1,405 @@
+            |(
+                path,
+                PerFilePendingFixes {
+                    mut file_contents,
+                    pending_fixes,
+                },
+            )| {
+                let mut violations: Vec<ViolationWithContext> = Default::default();
+                run_fixing_loop(
+                    &mut violations,
+                    &mut file_contents,
+                    pending_fixes,
+                    tree_sitter_grep_args.clone(),
+                    &aggregated_queries,
+                    &path,
+                    &config,
+                );
+                (path, (file_contents, violations))
+            },
+        )
+        .collect();
+    write_files(
+        aggregated_results_from_files_with_fixes
+            .iter()
+            .map(|(path, (file_contents, _))| (&**path, &**file_contents)),
+    );
+    for (path, (_, violations)) in aggregated_results_from_files_with_fixes {
+        all_violations.insert(path, violations);
+    }
+    all_violations.into_values().flatten().collect()
+}
+
+fn run_fixing_loop(
+    violations: &mut Vec<ViolationWithContext>,
+    file_contents: &mut Vec<u8>,
+    mut pending_fixes: HashMap<RuleName, Vec<PendingFix>>,
+    tree_sitter_grep_args: tree_sitter_grep::Args,
+    aggregated_queries: &AggregatedQueries,
+    path: &Path,
+    config: &Config,
+) {
+    for _ in 0..MAX_FIX_ITERATIONS {
+        apply_fixes(file_contents, pending_fixes);
+        pending_fixes = Default::default();
+        if config.report_fixed_violations {
+            *violations = violations
+                .iter()
+                .filter(|violation| violation.was_fix)
+                .cloned()
+                .collect();
+        } else {
+            violations.clear();
+        }
+        tree_sitter_grep::run_for_slice_with_callback(
+            file_contents,
+            tree_sitter_grep_args.clone(),
+            |capture_info| {
+                let (rule, rule_listener) =
+                    aggregated_queries.get_rule_and_listener(capture_info.pattern_index);
+                let mut query_match_context =
+                    QueryMatchContext::new(path, file_contents, rule, config);
+                (rule_listener.on_query_match)(capture_info.node, &mut query_match_context);
+                if let Some(reported_violations) = query_match_context.violations.take() {
+                    violations.extend(reported_violations);
+                }
+                if let Some(fixes) = query_match_context.into_pending_fixes() {
+                    pending_fixes
+                        .entry(rule.meta.name.clone())
+                        .or_default()
+                        .extend(fixes);
+                }
+            },
+        )
+        .unwrap();
+        if pending_fixes.is_empty() {
+            break;
+        }
+    }
+}
 
-pub fn run(args: Args) {
-    let language = tree_sitter_rust::language();
-    let context = Context::new(language);
```

Comment:
It was smelling like `Context` wanted to just be `Args` (aka `Config`) so I got rid of it